packages/machine/src/models/state-channel.ts

```diff
@@ -0,0 +1,239 @@
+      this.appInstances.entries()
+    );
+
+    const freeBalanceAppIndexes = new Map<AssetType, string>(
+      this.freeBalanceAppIndexes.entries()
+    );
+
+    appInstances.set(fb.id, fb);
+
+    freeBalanceAppIndexes.set(AssetType.ETH, fb.id);
+
+    return new StateChannel(
+      this.multisigAddress,
+      this.multisigOwners,
+      appInstances,
+      freeBalanceAppIndexes,
+      this.monotonicNumInstalledApps + 1
+    );
+  }
+
+  public setState(appInstanceId: string, state: object) {
+    const appInstance = this.getAppInstance(appInstanceId);
+
+    const appInstances = new Map<string, AppInstance>(
+      this.appInstances.entries()
+    );
+
+    const freeBalanceAppIndexes = new Map<AssetType, string>(
+      this.freeBalanceAppIndexes.entries()
+    );
+
+    appInstances.set(appInstanceId, appInstance.setState(state));
+
+    return new StateChannel(
+      this.multisigAddress,
+      this.multisigOwners,
+      appInstances,
+      freeBalanceAppIndexes,
+      this.monotonicNumInstalledApps
+    );
+  }
+
+  public installApp(
+    appInstance: AppInstance,
+    aliceBalanceDecrement: BigNumber,
+    bobBalanceDecrement: BigNumber
+  ) {
+    const fb = this.getFreeBalanceFor(AssetType.ETH);
+    const currentState = fb.state as ETHBucketAppState;
+
+    const aliceBalance = currentState.aliceBalance.sub(aliceBalanceDecrement);
+    const bobBalance = currentState.bobBalance.sub(bobBalanceDecrement);
+
+    if (aliceBalance.lt(Zero) || bobBalance.lt(Zero)) {
+      throw Error(INSUFFICIENT_FUNDS);
+    }
+
+    const appInstances = new Map<string, AppInstance>(
+      this.appInstances.entries()
+    );
+
+    const freeBalanceAppIndexes = new Map<AssetType, string>(
+      this.freeBalanceAppIndexes.entries()
+    );
+
+    appInstances
+      .set(appInstance.id, appInstance)
+      .set(fb.id, fb.setState({ ...currentState, aliceBalance, bobBalance }));
+
+    return new StateChannel(
+      this.multisigAddress,
+      this.multisigOwners,
+      appInstances,
+      freeBalanceAppIndexes,
+      this.monotonicNumInstalledApps + 1
+    );
+  }
+
+  public uninstallApp(
+    appInstanceId: string,
+    aliceBalanceIncrement: BigNumber,
+    bobBalanceIncrement: BigNumber
+  ) {
+    const fb = this.getFreeBalanceFor(AssetType.ETH);
+    const appToBeUninstalled = this.getAppInstance(appInstanceId);
+
+    const currentState = fb.state as ETHBucketAppState;
+
+    const aliceBalance = currentState.aliceBalance.sub(aliceBalanceIncrement);
+    const bobBalance = currentState.bobBalance.sub(bobBalanceIncrement);
+
+    const appInstances = new Map<string, AppInstance>(
+      this.appInstances.entries()
+    );
+
+    const freeBalanceAppIndexes = new Map<AssetType, string>(
+      this.freeBalanceAppIndexes.entries()
+    );
+
+    appInstances.delete(appToBeUninstalled.id);
+
+    appInstances.set(
+      fb.id,
+      fb.setState({ ...currentState, aliceBalance, bobBalance })
+    );
+
+    return new StateChannel(
```

Comment:
It doesn't feel right to me that we return new instances of `StateChannel` objects from each of these methods. It seems like the `StateChannel` object reference should be stable across all these operations.

Reply:
Do you mean that the class should not be immutable or that the implementation should hide the fact that it is immutable?

Reply:
@snario It depends. It could just be a naming issue. Are these objects purely internal to the machine or are they vended out to the Node / Playground Server as well?

Reply:
At the moment, purely internal. However, the intention is to use them in the `Node`. It is unclear how they will be distributed. My recommendation is to make the `machine` a subfolder of `Node` and hoist the `models` folder up into the `node` source code

Reply:
@snario Then I think they should probably be like how CF.js's `AppInstance` objects work in that state channels are singletons. API consumers only see one consistent state channel object.